Return Not found Ids in ActiveRecord::NotFound

gaurish opened the pull request:
### Summary

This builds on top of 15e2da656f41af0124f7577858536f3b65462ad5.
now it also returns exact Ids which were not found.

### Use-Case - APIs

For example: if Company API is expecting `params[:developer_ids]` as input param. The client sends say 20ids of out of 18 are valid & 2 are invalid. In current version of rails, client will get the error message as given below:

> Couldn't find all Developers with 'id': (7, 2, 19, 20, 10, 15, 1, 14, 8, 11, 12, 9, 6, 5, 17, 13, 18, 4, 16, 3) (found 18 results, but was looking for 20).

Now after reading the above message, it pretty hard to figure which IDs are invalid. But After this change, the error message will tell them exactly what's wrong:


> Couldn't find all Developers with 'id': (7, 2, 19, 20, 10, 15, 1, 14, 8, 11, 12, 9, 6, 5, 17, 13, 18, 4, 16, 3) (found 18 results, but was looking for 20). Please check following 'ids': 12,18.


Good Idea?

eileencodes commented on activerecord/lib/active_record/relation/finder_methods.rb:
```diff
@@ -342,8 +342,8 @@ def raise_record_not_found_exception!(ids = nil, result_size = nil, expected_siz
         error = "Couldn't find all #{name.pluralize} with '#{key}': ".dup
-        error << "(#{ids.join(", ")})#{conditions} (found #{result_size} results, but was looking for #{expected_size})"
-
+        error << "(#{ids.join(", ")})#{conditions} (found #{result_size} results, but was looking for #{expected_size}). "
```
Can you remove the extra space you added here?

eileencodes reviewed (changes requested):
Thanks for the PR. I'll merge once you remove the extra space noted in the review.

eileencodes commented:
Sorry I should have mentioned this before - can you squash your commits so there is only one? Thanks!

gaurish commented:
@eileencodes 
Thanks for taking the time to review. I have two questions:

1. There should a space in between sentences right, after "20)" & before "Please". correct? (found 18 results, but was looking for 20). Please check following 'ids': 12,18.
2. Does this require a changelog entry? Let me know I will add one.

matthewd commented:
"Please check" seems a bit casual to me.. can we perhaps better integrate the list with the existing message instead?

gaurish commented:
@matthewd if current mag isn't good then i can use whatever copy. For example we could say

found 18 results, but was looking for 20. Couldnt find Developers with id 12,18.

Let me know

gaurish commented:
btw, If no changes are needed.  @eileencodes has already reviewed & It can be merged.

kirs commented on activerecord/test/cases/associations/has_many_through_associations_test.rb:
```diff
@@ -889,7 +889,8 @@ def test_collection_singular_ids_setter_raises_exception_when_invalid_ids_set
     ids = [Developer.first.id, -9999]
     e = assert_raises(ActiveRecord::RecordNotFound) { company.developer_ids = ids }
-    assert_match(/Couldn't find all Developers with 'id'/, e.message)
+    msg = "Couldn't find all Developers with 'id': (1, -9999) (found 1 results, but was looking for 2).Please check following 'id': -9999."
```
Should there be a space between `for 2).Please`?

gaurish replied:
@kirs 
I added it in my original PR but @eileencodes asked me to remove the space.

eileencodes replied:
I meant there should be a space between sentences but not at the end of a sentence after a period.

eileencodes replied:
So the space should go at the beginning of the next sentence. Otherwise if there aren't found ID's you end up with:

"This is a sentence. " instead of "This is a sentence."

kirs commented:
> Couldn'tt find Developers with id 12,18.

Sounds better than "Please check" 👍

gaurish commented:
@eileencodes 
Updated. Ready for another pass

sgrif commented:
The error message still says "Please check"

eileencodes reviewed (changes requested):
Can you change the "Please check" copy to "Couldn't find Developers with id 12,18."?

gaurish committed:
Return Not found Ids in ActiveRecord::NotFound

This builds on top of 15e2da656f41af0124f7577858536f3b65462ad5.
now it also returns exact Ids which were not found which will be debugging simple.

gaurish commented:
@eileencodes Updated as per your feedback :)

gaurish commented:
@eileencodes bumping this incase this missed your attention last time
